// @ms-mwf/moray/src/index.js
export { default as Alert } from './components/alert/alert';
export { default as AutoComplete } from './components/autocomplete/autocomplete';
export { default as BackToTop } from './components/back-to-top/back-to-top';
export { default as Carousel } from './components/carousel/carousel';
export { default as CharacterCount } from './components/forms/character-count/character-count';
export { default as ClickGroup } from './components/click-group/click-group';
export { default as Collapse } from './components/collapse/collapse';
export { default as CollapseControls } from './components/collapse/collapse-controls';
export { default as ColorPicker } from './components/color-picker/color-picker';
export { default as ContentSwap } from './components/content-swap/content-swap';
export { default as Dropdown } from './components/dropdown/dropdown';
export { default as FormStar } from './components/forms/interactive-star-rating/interactive-star-rating';
export { default as FormValidation } from './components/forms/validation/validation';
export { default as Modal } from './components/modal/modal';
export { default as MultiFeature } from './components/multi-feature/multi-feature';
export { default as Popover } from './components/popover/popover';
export { default as Positioner } from './components/positioner/positioner';
export { default as Range } from './components/forms/range/range';
export { default as ShowMoreShowLess } from './components/show-more-show-less/show-more-show-less';
export { default as Sticky } from './components/sticky/sticky';
export { default as Tab } from './components/tab/tab';
export { default as TabSlider } from './components/tab/tab-slider';
export { default as Tooltip } from './components/tooltip/tooltip';

export { default as Util } from './js/util';
export { default as Debug } from './js/debug';
export { version } from '../package.json';

import initializeComponents, { MWF_INITIALIZED } from './components/initializeComponents';

document.addEventListener('DOMContentLoaded', () => {
  initializeComponents();

  document.dispatchEvent(new CustomEvent(MWF_INITIALIZED));
});
